### ISSUE TITLE
Adding a CompressedSplitSQLDataFrameList?

### ISSUE THREAD

**Issue by LTLA** [Bioconductor/SQLDataFrame#4]
Pretty much as the question says. It would be very nice to be able to do:

```r
# Using the 'obj' constructed in the vignette.
stuff <- splitAsList(obj, obj$division)
```

and then get something that respects the `CompressedSplitDataFrameList` contract. This would allow me to swap in a `SQLDataFrame` for a `DFrame` without having to change the rest of my code.

**Comment by LTLA** [comment]
Hold on. Despite its name, the `SQLDataFrame` is not actually a `DataFrame`. What's up with that? Never mind this request, I can't even use these objects inside an SE.

**Comment by rworkflow** [comment]
Hi @LTLA , the `SQLDataFrame` is more of a `DataFrame`-like object now,  It didn't extends the `DataFrame` directly. You are right on this! The slot of `tblData` holds the data as a `dbplyr::tbl_dbi` object. Although it supports some of the `DataFrame` methods, such as `[` and `[[` subsetting, the data is saved as a whole instead of a list of columns.

I recall that when we talk about implementing `SQLDataFrame` into a SE object, @hpages will need to redefine the `DataFrame` so that it works like a top level virtual class and doesn't have the restrictions such as the list property, and SE depends on this virtual `DataFrame` only. Then the `DFrame` and `SQLDataFrame` are second-level classes that extend this virtual class and can have different features. `DFrame` can still have the list property and `SQLDataFrame` to have a different set of design, so that `SQLDataFrame` can be eventually implemented in SE for very big annotation dataset. 

Sorry if I have mistaken anything here since some memories were getting vague. @hpages , could you kindly remind me the changes you have made on the `DataFrame` and `DFrame` concerning this? Thanks!

**Comment by Liubuntu** [comment]
Hi @LTLA , the `SQLDataFrame` is more of a `DataFrame`-like object now,  It didn't extends the `DataFrame` directly. You are right on this! The slot of `tblData` holds the data as a `dbplyr::tbl_dbi` object. Although it supports some of the `DataFrame` methods, such as `[` and `[[` subsetting, the data is saved as a whole instead of a list of columns.

I recall that when we talk about implementing `SQLDataFrame` into a SE object, @hpages will need to redefine the `DataFrame` so that it works like a top level virtual class and doesn't have the restrictions such as the list property, and SE depends on this virtual `DataFrame` only. Then the `DFrame` and `SQLDataFrame` are second-level classes that extend this virtual class and can have different features. `DFrame` can still have the list property and `SQLDataFrame` to have a different set of design, so that `SQLDataFrame` can be eventually implemented in SE for very big annotation dataset. 

Sorry if I have mistaken anything here since some memories were getting vague. @hpages , could you kindly remind me the changes you have made on the `DataFrame` and `DFrame` concerning this? Thanks!

**Comment by LTLA** [comment]
It would seem that we're pretty close. For example, you could imagine generalizing `DataFrame`'s `listData` so that it can be `ANY` object with the contract of supporting `length()` to return the number of available columns and `[[` with an integer index to return a specific vector of `nrows` length. Then `SQLDFrame` could subclass `DelayedDataFrame`, overriding the `listData` entry to require a SQL connection object instead.

**Comment by lawremi** [comment]
Still trying to understand the aversion to the `@listData` slot. We've already discussed the idea of putting high-level promise objects in it that refer to the individual columns.

**Comment by LTLA** [comment]
Wouldn't promises be pretty unstable? They'd get evaluated at the drop of a hat, e.g., if we serialized the object.

**Comment by lawremi** [comment]
That's true for R-level promises. I was referring to _high-level_ promises, i.e., promises implemented at the S4 level. That enables users, for example, to extract a column without forcing evaluation. The rsolr and hailr packages have explored this, fairly successfully.

**Comment by Liubuntu** [comment]
Hi @lawremi , I looked at our past emails about the `promise` implementation potentially in `DelayedDataFrame`. Still it looks promising to me for the big overhaul, however, as a very low-level implementation which may lead to many potential class extensions later, I would recommend some R/Bioconductor-core people who have more experiences and resources in getting help. I am not in the core now, and couldn't promise a good implementation in time...

**Comment by Liubuntu** [comment]
Hi @LTLA I am looking at the `DelayedDataFrame` and it extends the current `DFrame` with the `listData` slot, with an additional `lazyIndex` slot to save the mapping indexes for each column of the data. As a subclass, is it possible to generalize the `listData` slot here to be `ANY` object? I think the way should be to extend the virtual class (not implemented yet, right @hpages ?) of `DataFrame` and define a new slot for the data with `ANY` object?
